### PR TITLE
Logging, profiling and random seed

### PR DESCRIPTION
**Pull request by iainrb**
Logging and profiling improvements:
- Run with --profile to write Python profiler output to STDOUT
- Option --log sets log path, otherwise logs to STDERR
- Options --verbose and --debug set log level
- Option to set alternative random seed, with updated tests

### CODE REVIEW

File: bam_qc_metrics/bam_qc.py [hunk 1]
```diff
@@ -2,9 +2,9 @@
 
 """Main class to compute BAM QC metrics"""
 
-import bam_qc_metrics, csv, json, os, re, pybedtools, pysam, sys, tempfile
+import bam_qc_metrics, csv, json, logging, math, os, re, pybedtools, pysam, sys, tempfile
 
-class base:
+class base(object):
```
Comment (slazicoicr) [inline bam_qc_metrics/bam_qc.py]: No need to inherit from `object` in Python 3, but does not hurt.

File: bam_qc_metrics/bam_qc.py [hunk 2]
```diff
@@ -101,50 +111,70 @@ def __init__(self, config):
...
                                          self.reference,
                                          self.expected_insert_max,
-                                         self.n_as_mismatch)
+                                         self.n_as_mismatch,
+                                         self.logger)
         self.fast_metrics = self.update_unmapped_count(fast_finder.metrics)
+        self.logger.info("Finished computing fast bam_qc metrics")
         # apply downsampling (if any)
         if self.sample_rate != None and self.sample_rate > 1:
             slow_finder_input_path = self.generate_downsampled_bam(fast_finder_input_path,
```
Comment (slazicoicr) [inline bam_qc_metrics/bam_qc.py]: As you are logging when no down-sampling occurs, also log what the down sample rate is when it does occur.

File: bam_qc_metrics/bam_qc.py [hunk 3]
```diff
@@ -158,33 +188,66 @@ def cleanup(self):
...
-                             "temporary directory %s\n" % self.tmpdir)
+            self.logger.info("Cleaned up default temporary directory %s" % self.tmpdir)
+        else:
+            self.logger.info("Omitting cleanup for user-specified "+\
+                             "temporary directory %s" % self.tmpdir)
 
+    def configure_logger(self, log_path=None):
+        logger = logging.getLogger(__name__)
+        if self.debug:
+            log_level = getattr(logging, 'DEBUG')
```
Comment (slazicoicr) [inline bam_qc_metrics/bam_qc.py]: Why not `logging.DEBUG`?

File: bam_qc_metrics/bam_qc.py (hunk 2 above)
Comment (iainrb) [reply bam_qc_metrics/bam_qc.py]: Already being logged here: https://github.com/oicr-gsi/bam-qc-metrics/blob/29880cf5b6556042d6e5835d6fbf488f4199f05a/bam_qc_metrics/bam_qc.py#L233

File: bam_qc_metrics/bam_qc.py (hunk 3 above)
Comment (iainrb) [reply bam_qc_metrics/bam_qc.py]: No particular reason. `logging.DEBUG` is slightly safer, will fix.

**Pull request merged**